persist/fs/commitlog/reader.go

```diff
@@ -109,58 +192,267 @@ func (r *reader) Read() (
+	// 			[0]: [a, c]
+	// 			[1]: [b, d]
+	// If we now read from the outBufs in round-robin order (0, 1, 0, 1) we will
+	// end up with the correct global ordering: a, b, c, d
+	rr, ok := <-r.outBufs[r.nextIndex%r.numConc]
+	if !ok {
+		return Series{}, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), io.EOF
+	}
+	if rr.pendingMetadata != nil {
+		// Wait for metadata to be available
+		rr.series, rr.resultErr = rr.pendingMetadata.wait()
 	}
+	r.nextIndex++
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok && len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) startBackgroundWorkers() error {
+	// Make sure background workers are never setup more than once
+	set := atomic.CompareAndSwapInt64(&r.bgWorkersInitialized, 0, 1)
+	if !set {
+		return errCommitLogReaderMultipleReadloops
+	}
+
+	// Start background worker goroutines
+	go r.readLoop()
+	for i, decoderBuf := range r.decoderBufs {
+		localDecoderBuf, outBuf := decoderBuf, r.outBufs[i]
+		go r.decoderLoop(localDecoderBuf, outBuf)
+	}
+
+	return nil
+}
+
+func (r *reader) readLoop() {
+	index := 0
+	defer r.shutdown()
+	for {
+		select {
+		case <-r.cancelCtx.Done():
 			return
+		default:
+			data, err := r.readChunk()
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
+			if err == io.EOF {
+				return
+			}
+		}
+	}
+}
+
+func (r *reader) shutdown() {
+	for _, decoderBuf := range r.decoderBufs {
+		close(decoderBuf)
+	}
+	r.shutdownCh <- r.close()
+}
+
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			r.writeToOutBuf(outBuf, readResponse)
+			continue
 		}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		metadata = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       decoded.Shard,
+		arg.bytes.IncRef()
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			r.writeToOutBuf(outBuf, readResponse)
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
+		if len(entry.Metadata) != 0 {
```

Comment:
can you flip the check and reduce the nesting here? something like:

```
if len(entry.Metadata) == 0 {
  continue 
}

// rest of the code
```

Reply:
I don't think so because there is more code that needs to execute afterwards regardless